### PR TITLE
CMake improvements

### PR DESCRIPTION
**Pull request by rbberger**
## Purpose

This PR implements some of the open issues mentioned in #616.

* cleans up any style headers which might be in the `src` folder from previous compilations
* ignores package headers and source files which were installed by copying them into the `src` folder.

## Author(s)

Richard Berger (Temple U)

## Post Submission Checklist

_Please check the fields below as they are completed_
- [x] The feature or features in this pull request is complete
- [x] The source code follows the LAMMPS formatting guidelines

### CODE REVIEW

File: cmake/CMakeLists.txt [hunk 1]
```diff
@@ -330,13 +328,32 @@ RegisterStyles(${LAMMPS_SOURCE_DIR})
...
+  set(${PKG}_SOURCES_DIR ${LAMMPS_SOURCE_DIR}/${PKG})
+
+  # ignore PKG files which were manually installed in src folder
+  # headers are ignored during RegisterStyles
+  file(GLOB ${PKG}_SOURCES ${${PKG}_SOURCES_DIR}/*.cpp)
+  file(GLOB ${PKG}_HEADERS ${${PKG}_SOURCES_DIR}/*.h)
+
+  foreach(PKG_FILE in ${${PKG}_SOURCES})
+      get_filename_component(FNAME ${PKG_FILE} NAME)
+      list(FIND LIB_SOURCES ${LAMMPS_SOURCE_DIR}/${FNAME} IDX)
```
Comment (junghans) [inline cmake/CMakeLists.txt]: Why not `list(REMOVE_ITEM ..)`?

File: cmake/CMakeLists.txt [hunk 2]
```diff
@@ -330,13 +328,32 @@ RegisterStyles(${LAMMPS_SOURCE_DIR})
...
 
+      if(NOT (IDX LESS 0))
+          #message("Ignoring ${PKG_FILE}...")
+          list(REMOVE_AT LIB_SOURCES ${IDX})
+      endif()
+  endforeach()
+
+  foreach(PKG_FILE in ${${PKG}_HEADERS})
+      get_filename_component(FNAME ${PKG_FILE} NAME)
+      DetectAndRemovePackageHeader(${LAMMPS_SOURCE_DIR}/${FNAME})
```
Comment (junghans) [inline cmake/CMakeLists.txt]: Will this solve the problem of `USER-MISC` always depending on `MANYBODY`?

**Comment by junghans** [comment]
> Using `list(FIND ..)` and `if(NOT (... LESS_THAN 0))` for better backward compatibility with older CMake versions.

We require CMake-3.1, but which version are you aiming for with backward compatibility?

File: cmake/CMakeLists.txt (hunk 1 above)
Comment (rbberger) [inline cmake/CMakeLists.txt]: "I didn't see the forest due to all the trees" :wink:

**Comment by rbberger** [comment]
@junghans originally I was looking into using `IN_LIST` and `GREATHER_EQUAL`, which are only available in later versions. By simplifying that all goes away. Thanks for the review!

File: cmake/CMakeLists.txt (hunk 2 above)
Comment (rbberger) [inline cmake/CMakeLists.txt]: No, I didn't look into that. That is a per-file dependency which is a different problem.

File: cmake/CMakeLists.txt [hunk 3]
```diff
@@ -23,13 +23,11 @@ if(NOT CMAKE_BUILD_TYPE AND NOT CMAKE_CXX_FLAGS)
...
-             style_fix.h style_improper.h style_integrate.h style_kspace.h style_minimize.h style_nbin.h style_npair.h style_nstencil.h
-             style_ntopo.h style_pair.h style_reader.h style_region.h)
-  if(EXISTS ${LAMMPS_SOURCE_DIR}/${STYLE_FILE})
-    message(FATAL_ERROR "There is a ${STYLE_FILE} in ${LAMMPS_SOURCE_DIR}, please clean up the source directory first")
-  endif()
-endforeach()
+# remove any style headers in the src dir
+file(GLOB SRC_STYLE_FILES ${LAMMPS_SOURCE_DIR}/style_*.h)
+if(SRC_STYLE_FILES)
+  file(REMOVE ${SRC_STYLE_FILES})
```
Comment (junghans) [inline cmake/CMakeLists.txt]: Isn't that dangerous? If I name a file `style_wall_cj.h` or something?

Comment (rbberger) [inline cmake/CMakeLists.txt]: No, it's "enforcing" coding standards. Creating a file with such a name is asking for it. I would simply say `style_` is reserved for the build system. So all it does is nuking files that shouldn't start with `style_` in the first place.

Comment (junghans) [inline cmake/CMakeLists.txt]: Ok, if it is the same in make, I am ok it that.

**Comment by rbberger** [comment]
Moving all approved CMake PRs into #628 and closing this one.

**Pull request closed without merging**